shard/sharder.go

```diff
@@ -0,0 +1,352 @@
+// that tracks this node's graph
+func (s *Sharder) AddNode(node ipld.Node) error {
+	logger.Debug("adding node to shard")
+	logger.Debugf("sharder size: %d---sharder thresh: %d", s.byteCount, s.byteThreshold)
+	size, err := node.Size()
+	if err != nil {
+		return err
+	}
+	c := node.Cid()
+	format, ok := cid.CodecToStr[c.Type()]
+	if !ok {
+		format = ""
+		logger.Warning("unsupported cid type, treating as v0")
+	}
+	if c.Prefix().Version == 0 {
+		format = "v0"
+	}
+	if s.unInit() {
+		logger.Debug("initializing next shard of data")
+		if err := s.initShard(); err != nil {
+			return err
+		}
+		logger.Debugf("After first init. thresh: %d", s.byteThreshold)
+	} else {
+		if s.byteCount+size+s.clusterDAGCountBytes() > s.byteThreshold {
+			logger.Debug("shard at capacity, pin cluster DAG node")
+			if err := s.Flush(); err != nil {
+				return err
+			}
+			if err := s.initShard(); err != nil {
+				return err
+			}
+			logger.Debugf("After flushing. thresh: %d", s.byteThreshold)
+		}
+	}
+
+	// Shard is initialized and can accommodate node by config-enforced
+	// invariant that shard size is always greater than the ipfs block
+	// max chunk size
+	logger.Debugf("Adding size: %d to byteCount at: %d", size, s.byteCount)
+	s.byteCount += size
+
+	key := fmt.Sprintf("%d", len(s.currentShard))
+	s.currentShard[key] = node.Cid()
+	var retStr string
+	b := api.BlockWithFormat{
+		Data:   node.RawData(),
+		Format: format,
+	}
+	return s.rpcClient.Call(s.assignedPeer, "Cluster", "IPFSBlockPut",
+		b, &retStr)
+}
+
+// Flush completes the allocation of the current shard of a file by adding the
+// clusterDAG shard node block to IPFS.  It clears sharder state in order to
+// clear the cluster sharding component to shard new files in the case the last
+// shard is being flushed.
+func (s *Sharder) Flush() error {
+	// Serialize shard node and reset state
+	logger.Debugf("Flushing the current shard %v", s.currentShard)
+	shardNode, err := cbor.WrapObject(s.currentShard, mh.SHA2_256, mh.DefaultLengths[mh.SHA2_256])
+	if err != nil {
+		return err
+	}
+	logger.Debugf("The dag cbor Node Links: %v", shardNode.Links())
+
+	targetPeer := s.assignedPeer
+	s.currentShard = make(map[string]*cid.Cid)
+	s.assignedPeer = peer.ID("")
+	s.byteThreshold = 0
+	s.byteCount = 0
+	var retStr string
+	b := api.BlockWithFormat{
+		Data:   shardNode.RawData(),
+		Format: "cbor",
+	}
+	logger.Debugf("Here is the serialized ipld: %x", b.Data)
+	err = s.rpcClient.Call(targetPeer, "Cluster", "IPFSBlockPut",
+		b, &retStr)
+	if err != nil {
+		return err
+	}
+
+	// Track shard node as a normal pin within cluster state
+	pinS := api.Pin{
+		Cid:                  shardNode.Cid(),
+		Allocations:          []peer.ID{targetPeer},
+		ReplicationFactorMin: 1,
+		ReplicationFactorMax: 1,
+	}.ToSerial()
+
+	return s.rpcClient.Call(targetPeer, "Cluster", "Pin", pinS, &struct{}{})
+}
+
+/* Open questions
+
+1. getAssignment()
```

Comment:
I might be underthinking this.

1. For every new shard, get a node from the allocator (if using the `reposize` strategy, the one reporting less reposize).
2. IPFS Block put all the blocks to that node
3. IPFS Block put the shard root node to that node
4. ipfs-cluster-pin the root node like you would pin any other CID (so it will be pinned recursively in several cluster peers)

The last step will pin with the desired replication factor just like any other CID, and will tell the necessary number of ipfs daemons to pin that tree. IPFS daemons which don't have the content will fetch it from the ipfs peer that has it.

You don't need to think how much space you take from a peer. It's going to be the shard size because each shard is written to a single place at first. For our prototype, we don't even have to worry about making this first peer the same as the ones in which the shard is pinned. You may well write to one peer and then the `Pin` operation may allocate them to others because it has no context that that shard is already written in one of them (it should be easy to address this later and manually allocate that peer).

Don't worry now about shards not distributed evenly etc. we can address not-up-to-date metrics, more clever allocation assignments later.

The final root node of all shards needs to be pinned, perhaps everywhere, but not recursively.

Reply:
Well said.  I wrote this before implementing and I converged pretty much onto the implementation you describe in 1 to 4.  

I understand that the space you take from a peer is just the shard size, in this section I was focusing (in retrospect looks like too much) on the details of the shard size potentially adapting during the sharding process.  Looking back most of the concerns here are orthogonal to sharding, i.e. how can nodes specify exactly what the storage they offer up can be used for: do they have the bandwidth for big shards, which nodes are make sense to accept allocs from for bandwidth or region/topology considerations, do they want to maintain extra freespace for more small pins in the future?

Ignoring the above as out of scope and taking on the simplifying assumption that all nodes in a cluster agree on one shard size, and it's a configured value, most of the allocation complexity goes away.  The only thing left to handle is what happens in the edge case all nodes together have the space to fit file data but no node has the space for the configured shard size (need fewer smaller shards).

Eventually it may make sense to stop making this assumption and allow nodes to specify richer inputs into the allocation process.  This will probably happen to address the broader questions asked in paragraph 2.  In the same way the reposize metric gets used for allocs perhaps the allocation configs of individual nodes will broadcast their allocation preferences, including their max shard size.  One weird thing about the current assumption for sharding is that as it stands right now it is ONLY the sharder config of the node doing the `ipfs-cluster-ctl add` that decides the shard size.  Therefore a node could specify a shard size of 5 MB in its config and end up with 100MB shards pinned to its repo.  

Also thanks for pointing out that we'd need to change pinning to include a particular PID to make the pin happen on the inserted node.  Right now we are doing step 4 but I forgot that this doesn't necessarily mean that the node with the blocks will do the pinning.  In fact there might be something weird happening where the node given the blocks is less likely to get allocated the pin because the freespace in its repo is artificially lowered by the inserted blocks.  Also supporting higher repl factors becomes pretty hard in the case of each node specifying their own shard size because then the same file may need to be sharded into different sized pieces to pack it into different nodes.

Whew sorry about the essay, I think we are on the same page, just wanted to get those thoughts out in the open.

Reply:
Forgot to submit this yesterday:

> Therefore a node could specify a shard size of 5 MB in its config and end up with 100MB shards pinned to its repo

Not only here, in many places we make the assumptions that cluster configurations are similar and make sense for the cluster. Shard size should be a request option anyway (for a start), and whoever triggers a sharding-add should be careful to not do it in a way that is not going to fit the cluster (I think that's a respectable expectation for our first approach, and that these considerations can be addressed in later rounds).

> Another thought: supporting higher repl factors becomes pretty hard in the case of each node specifying their own shard size because then the same file may need to be sharded into different sized pieces to pack it into different nodes. 

Having different shards sizes depending on available space might be a really good idea with things like FEC. But yeah, it's weird on a >1 replication factor. We will have to think of this, but probably the diverse-cluster with nodes of many types with completely different repo-capacity and almost full is not the main target usecase for our first iteration.